google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1beta2/StreamWriterV2Test.java

```diff
@@ -0,0 +1,293 @@
+
+    ApiFuture<AppendRowsResponse> appendFuture2 = sendTestMessage(writer, new String[] {"B"});
+    assertEquals(1, appendFuture2.get().getAppendResult().getOffset().getValue());
+
+    writer.close();
+  }
+
+  @Test
+  public void testAppendAfterUserClose() throws Exception {
+    StreamWriterV2 writer = getTestStreamWriterV2();
+    testBigQueryWrite.addResponse(createAppendResponse(0));
+
+    ApiFuture<AppendRowsResponse> appendFuture1 = sendTestMessage(writer, new String[] {"A"});
+    writer.close();
+    ApiFuture<AppendRowsResponse> appendFuture2 = sendTestMessage(writer, new String[] {"B"});
+
+    assertEquals(0, appendFuture1.get().getAppendResult().getOffset().getValue());
+    assertTrue(appendFuture2.isDone());
+    StatusRuntimeException actualError =
+        assertFutureException(StatusRuntimeException.class, appendFuture2);
+    assertEquals(Status.Code.FAILED_PRECONDITION, actualError.getStatus().getCode());
+  }
+
+  @Test
+  public void testAppendAfterServerClose() throws Exception {
+    StreamWriterV2 writer = getTestStreamWriterV2();
+    testBigQueryWrite.addException(Status.INTERNAL.asException());
+
+    ApiFuture<AppendRowsResponse> appendFuture1 = sendTestMessage(writer, new String[] {"A"});
+    ApiException error1 = assertFutureException(ApiException.class, appendFuture1);
+    assertEquals(Code.INTERNAL, error1.getStatusCode().getCode());
+
+    ApiFuture<AppendRowsResponse> appendFuture2 = sendTestMessage(writer, new String[] {"B"});
+    assertTrue(appendFuture2.isDone());
+    StatusRuntimeException error2 =
+        assertFutureException(StatusRuntimeException.class, appendFuture2);
+    assertEquals(Status.Code.FAILED_PRECONDITION, error2.getStatus().getCode());
+
+    writer.close();
+  }
+
+  @Test
+  public void userCloseWhileRequestInflight() throws Exception {
+    final StreamWriterV2 writer = getTestStreamWriterV2();
+    // Server will sleep 2 seconds before sending back the response.
+    testBigQueryWrite.setResponseSleep(Duration.ofSeconds(2));
+    testBigQueryWrite.addResponse(createAppendResponse(0));
+
+    // Send a request and close the stream in separate thread while the request is inflight.
+    final ApiFuture<AppendRowsResponse> appendFuture1 = sendTestMessage(writer, new String[] {"A"});
+    Thread closeThread =
+        new Thread(
+            new Runnable() {
+              @Override
+              public void run() {
+                writer.close();
+              }
+            });
+    closeThread.start();
+
+    // Due to the sleep on server, the append won't finish within 1 second even though stream
+    // is being closed.
+    assertThrows(
+        TimeoutException.class,
+        new ThrowingRunnable() {
+          @Override
+          public void run() throws Throwable {
+            appendFuture1.get(1, TimeUnit.SECONDS);
+          }
+        });
+
+    // Within 2 seconds, the request should be done and stream should be closed.
+    closeThread.join(2000);
```

Comment:
I think these ways can be replaced by fake clock. The test code on the old streamwriter should work.

Reply:
Yeah, I would like to do that as well. But I cannot find a fake clock implementation in Java, that supports sleep.

Reply:
The FakeBigQueryImpl already supports the fakeclock and adding these changes would work:
https://screenshot.googleplex.com/4hPUAWG2HMctcE7

Reply:
https://screenshot.googleplex.com/6wcobhWsBJggWvV